src/Compiler/CompilerUtils.cpp

```diff
@@ -308,6 +308,21 @@ static void genLLVMBitcode(const mlir::OwningOpRef<ModuleOp> &module,
   llvm::Metadata *identNode[] = {llvm::MDString::get(ctx, OnnxMlirVersion)};
   identMetadata->addOperand(llvm::MDNode::get(ctx, identNode));
 
+  // Annotate functions to be accessible from DLL on Windows.
+#ifdef _WIN32
+  SmallVector<StringRef, 4> exportedFuncs;
+  // TODO: support multiple entry points.
+  exportedFuncs.emplace_back(StringRef("run_main_graph"));
+  exportedFuncs.emplace_back(StringRef("omQueryEntryPoints"));
+  exportedFuncs.emplace_back(StringRef("omInputSignature"));
+  exportedFuncs.emplace_back(StringRef("omOutputSignature"));
+  for (StringRef funcName : exportedFuncs) {
+    llvm::GlobalValue *GV = llvmModule->getNamedValue(funcName);
+    GV->setDSOLocal(true);
+    GV->setDLLStorageClass(llvm::GlobalValue::DLLExportStorageClass);
+  }
+#endif
+
```

Comment:
@tungld Can you explain a bit how this worked? I would have expected that once you annotate the bitcode properly for `run_main_graph` etc. to be exported, the rest are all taken care of by llc and the linker. Why do we still have to do something in `CompilerUtils.cpp` to export the symbols?

Reply:
@gongsu832 Firstly we should distinguish "LLVMIR in mlir" which is a dialect in MLIR and "LLVMIR in llvm" which is the original LLVM existed long time ago before MLIR.

"LLVMIR in mlir" is limited and still not equivalent to "LLVMIR in llvm" , for example we cannot annotate a function with dllexport in "LLVMIR in mlir".

Now, our onnx-mlir  or CompilerUtils in particular, invokes KrnlToLLVM to generates "LLVMIR in mlir", then calls `translateModuleToLLVMIR` to translate "LLVMIR in mlir" to "LLVMIR in llvm`, then tranlates to bitcode, and link etc.

What I did here is to annotate functions of the generated code when they are in "LLVMIR in llvm" so that the functions are equivalent to having `__declspec(dllexport)` (e.g. when -emit-llvm for a C++ code with  `__declspec(dllexport)`. The rest will be all taken care of by llc and the linker as you said.

Reply:
Thanks @tungld ! Can you put this as comments above this block of code? Makes it easier to understand how this works.

Reply:
BTW, just saw your comment above:

> This patch cannot handle a model with multiple entry points, e.g. entry points are not `run_main_graph`, but something else that must be queried via `omQueryEntryPoints`.

Is this still true? There is no way to annotate entry points other than `run_main_graph`?

Reply:
It's feasible, e.g. entry points are constants in LLVMIR and we can read them. I will address this in another PR. It requires a little more effort.

Reply:
> Can you put this as comments above this block of code? Makes it easier to understand how this works.

I created a separate function `tailorLLVMIR` and put all related code there. Hope it make things clearer.

Reply:
> It's feasible, e.g. entry points are constants in LLVMIR and we can read them. I will address this in another PR. It requires a little more effort.

That would be perfect! With this we would then completely address the symbol export issue for Windows without a .def file.

Reply:
@gongsu832 anyway, I just added support for multiple entry points. So this patch fully supports dllexport for the generated code.